### ISSUE TITLE
Changed File Close Dialogue not usable with keyboard

### ISSUE THREAD

**Issue by ccoenen** [atom/atom#2928]
(this is possibly windows only, tested with 0.113.0)

**Steps To Reproduce**
- Open Atom
- Type something into a new editor or change an existing file (doesn't matter which)
- Attempt to close the tab (for example `Ctrl+W`)
- Correctly, a dialogue opens, that will give you a choice how to proceed

![close-dialog-keyboard-lacking](https://cloud.githubusercontent.com/assets/124909/3550749/38cadc72-08dd-11e4-8d71-83e665a150cd.png)

**current (faulty) behaviour**

The Keyboard is pretty much useless for this dialogue. You can only Hit `[Enter]` to save.

**Desired Behaviour**

These interactions all don't work, but would be very much needed, all of these are very much standard behaviour for any application:
- [X] `[left]` and `[right]` should switch focus between buttos (so we can select a different thing with keyboard only)
- [X] `[esc]` should Cancel
- [ ] `[alt]` should bring up shortcut labels (much like it does for the application menu, i.e. _F_ile, which means press `[Alt+F]`)
- [X] `[tab]` should cycle between the buttons, `[shift-tab]` should cycle backwards.

**Comment by oh-ren** [comment]
Hmm.. Regarding the `[alt]` key, that must be something of Windows > 7?
On Windows 7 I get dialogs where you instantly can press `Y` or `N` (no need for `[alt]`), and it's annoying the Atom close dialog doesn't support this. Yes, I can do tab + enter, but it's an extra action one has to make, and thus making it more error prone.. 
I see @zcbenz mentioning it's not possible to support `[alt]`, but is it possible to support `Y`/`N`?

**Comment by ccoenen** [comment]
they used to look like this:

![Old Dialog](https://cloud.githubusercontent.com/assets/124909/3890417/1ca28590-221e-11e4-8fd9-d13d80c02604.png)

and now they look like this: 

![delete](https://cloud.githubusercontent.com/assets/124909/10375617/e5b4fbc0-6df9-11e5-8307-865b2aa352aa.png)

The suggested <kbd>alt</kbd> behaviour is still not present. It is not essential and the other methods work fine (as indicated by the ticked boxes in the original post), but the "alt for shortcut" is a nice addition that is common on windows and other systems.

**Comment by oh-ren** [comment]
@ccoenen I've not used Windows beyond 7, but do later Windows version require you to first press <kbd>Alt</kbd> before you can use the hotkey (e.g. <kbd>Y</kbd>/<kbd>N</kbd>) in dialog windows?

**Comment by ccoenen** [comment]
No.

**Comment by itsravenous** [comment]
Just to say it's not just Windows, <kbd>Alt</kbd> + letter for dialogs is common on Linux too - I really miss this functionality (tabbing twice to the "don't save" button, my most common use case, is getting old)

**Comment by peppy** [comment]
Need Y/N key support badly. Currently a very disturbing dialog box which destroys flow on windows.

**Comment by tkothe** [comment]
Same issue on Mac. Or even worse: I can't even use left/right keys or tab to switch do the "don't save" button.
Do you want me to raise this as a separate issue or is this actually nothing that would turn out to be OS specific? Thanks!

**Comment by manuelbieh** [comment]
Same problem on Windows 10. Impossible to press Y/N to save (or not save) the file.

**Comment by bakert** [comment]
On Mac Esc works for Cancel and Enter works to Save. There is no way to indicate "Don't Save" from the keyboard and arrow keys/Tab don't work for moving around the dialog.

<img width="472" alt="screen shot 2016-04-30 at 3 00 14 pm" src="https://cloud.githubusercontent.com/assets/1541/14936439/bd320e22-0ee4-11e6-871b-7aa8b7e39e47.png">

If you chose to save an unnamed file the keyboard behavior is kind of weird there too. I can tab between Save As and Tags inputs but not to Where/Cancel/Save. Enter and Esc work fine here for Save and Cancel.

<img width="437" alt="screen shot 2016-04-30 at 3 02 54 pm" src="https://cloud.githubusercontent.com/assets/1541/14936441/c7823b36-0ee4-11e6-8c8f-1905763c86f8.png">

OSX 10.10.5 Yosemite. Atom 1.6.2.

**Comment by lee-dohm** [comment]
@bakert Those are standard OS X dialog boxes, they don't support many keyboard operations like you're describing unless you enable them in the Keyboard System Settings panel:

<img width="780" alt="screen_shot_2016-04-30_at_10_53_18_pm" src="https://cloud.githubusercontent.com/assets/1038121/14940313/77c74f7c-0f26-11e6-8427-263217f2859b.png">

**Comment by bakert** [comment]
Doh! Thanks for the heads up!

**Comment by damieng** [comment]
The situation here as far as I understand it is on macOS dialogs can not normally be navigated without turning on accessibility and that works already today.

Windows users can now navigate with those arrows as well.

There is some expectation for keyboard shortcuts/accelerators often represented by an underline in the button name. e.g. pressing <kbd>D</kbd> for `Don't save`

Electron doesn't specifically support those accelerators today likely because macOS doesn't and it started there.

Windows and Linux internally use the ampersand character in their low level APIs to make the following letter an accelerator.  e.g. `&Don't Save`

You can actually do this today and it works on Win/Linux but then shows as an actual & on macOS.

Ideally we'd expand Electron to support objects as well as strings in the button array parameter which could specify label and shortcut separately.  It would escape the label to remove any & and then apply the shortcut by matching the first character and putting a & in front of it.

```dialog.showMessageBox({message: 'Your changes will be lost unless you save them.', buttons: [ { label: 'Don't save', shortcut: 'D' }, 'Save' ]})```

If we were instead to just work around this we'd need to wrap our calls to the Electron dialog api with something that removed the accelerator on macOS.

**Comment by damieng** [comment]
Have create an issue over on the Electron repro to track the ability to use accelerators across platforms without an abstraction layer; https://github.com/electron/electron/issues/7517

**Comment by bronson** [comment]
The "Don't Save" shortcut on the Mac is `cmd-delete`: https://support.apple.com/en-us/HT201236

I've confirmed that cmd-delete works on MacVim, Text Edit, Sublime, Graph Sketcher, and some CAD programs.  The only one that it doesn't work on is Atom.

It seems like Atom must be accidentally disabling it, doesn't it?  (if so, should I open a new issue to track just this problem?)

**Comment by 50Wliu** [comment]
@bronson This is a Linux/Windows-only issue.  The macOS issue you are referencing has just been fixed (#13071).

**Comment by rubin110** [comment]
Does this ticket include the desire to have a shortcut for the action "Don't Save" on Linux? This is one of my few annoyances with Atom at the moment.

**Comment by damieng** [comment]
We've been waiting for shortcut accelerators in dialogs to land in Electron - they did in that in 1.4.14 https://github.com/electron/electron/pull/8308 so we need to upgrade Electron before we can do this.  That upgrade is being tracked in https://github.com/atom/atom/pull/12696

**Comment by Aetf** [comment]
Any updates on this? Since the mentioned pull requests have already been merged?

**Comment by damieng** [comment]
The PR to get the updated Electron has only landed on master. I just checked in the dialog fixes there. It should make the next beta and then hit stable the release after that.
